From 8cb046011aa87da4a7d9080455cc3d5f2ce59f71 Mon Sep 17 00:00:00 2001
From: Meticulus <theonejohnnyd@gmail.com>
Date: Mon, 18 Dec 2017 08:37:44 -0600
Subject: [PATCH] 
 "0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
 _______________________________

*Previous From 3fd182f9b2c56ea0547546391cae2b3e0f2cf7e4 Mon Sep 17 00:00:00 2001
*Previous From: Meticulus <theonejohnnyd@gmail.com>
*Previous Date: Mon, 18 Dec 2017 08:18:40 -0600
*Previous Subject: [PATCH]
*"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
*_______________________________
*
**Previous From d5c9198af68110ecdcb24c26a4884eb69e77018f Mon Sep 17 00:00:00 2001
**Previous From: Meticulus <theonejohnnyd@gmail.com>
**Previous Date: Sun, 17 Dec 2017 15:07:04 -0600
**Previous Subject: [PATCH]
**"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
**_______________________________
**
***Previous From 0ae26a6ee90ab53a717487034c2b93151aab61d3 Mon Sep 17 00:00:00 2001
***Previous From: Meticulus <theonejohnnyd@gmail.com>
***Previous Date: Sun, 17 Dec 2017 15:02:07 -0600
***Previous Subject: [PATCH]
***"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
***_______________________________
***
****Previous From b2ee57fed0116a031d7827c369af62f1214651d1 Mon Sep 17 00:00:00 2001
****Previous From: Meticulus <theonejohnnyd@gmail.com>
****Previous Date: Sun, 17 Dec 2017 14:57:58 -0600
****Previous Subject: [PATCH]
****"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
****_______________________________
****
*****Previous From 0d2c7f53d3b81a4b576e188cd493350eff6d7e62 Mon Sep 17 00:00:00 2001
*****Previous From: Meticulus <theonejohnnyd@gmail.com>
*****Previous Date: Sun, 17 Dec 2017 14:54:48 -0600
*****Previous Subject: [PATCH]
*****"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
*****_______________________________
*****
******Previous From bfe8ba6c3098647cbbb3c3f6e0c0303ea63bfc49 Mon Sep 17 00:00:00 2001
******Previous From: Meticulus <theonejohnnyd@gmail.com>
******Previous Date: Sun, 17 Dec 2017 14:41:29 -0600
******Previous Subject: [PATCH]
******"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
******_______________________________
******
*******Previous From b71f9a872af5e24e80d8fe4367ffe7170e2d474a Mon Sep 17 00:00:00 2001
*******Previous From: Meticulus <theonejohnnyd@gmail.com>
*******Previous Date: Sat, 16 Dec 2017 15:59:25 -0600
*******Previous Subject: [PATCH]
*******"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
*******_______________________________
*******
********Previous From b46b35e2b863f853a933c365d20d4889d3383492 Mon Sep 17 00:00:00 2001
********Previous From: Meticulus <theonejohnnyd@gmail.com>
********Previous Date: Tue, 12 Dec 2017 05:41:34 -0600
********Previous Subject: [PATCH]
********"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
********_______________________________
********
*********Previous From c3dd6d4148b3710b8b472f5b18c81680c4bd29c8 Mon Sep 17 00:00:00 2001
*********Previous From: Meticulus <theonejohnnyd@gmail.com>
*********Previous Date: Mon, 13 Nov 2017 11:04:12 -0600
*********Previous Subject: [PATCH]
*********"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
*********_______________________________
*********
**********Previous From 14b890bcf49207871cfc6b1a73a1304f39fc92fb Mon Sep 17 00:00:00 2001
**********Previous From: Meticulus <theonejohnnyd@gmail.com>
**********Previous Date: Fri, 10 Nov 2017 03:38:01 -0600
**********Previous Subject: [PATCH]
**********"0001-linker-Allow-shim-libs-with-enforcing-sepolicy-boot-critical
**********_______________________________
**********
***********Previous From a883e096f5e679976a71ff5e16f39b3a16239939 Mon Sep 17 00:00:00 2001
***********Previous From: Meticulus <theonejohnnyd@gmail.com>
***********Previous Date: Sat, 24 Dec 2016 23:49:04 -0600
***********Previous Subject: [PATCH] 0001-linker-Allow-shim-libs-with-enforcing-sepolicy
***********
***********Change-Id: I09ecbdfe092879e0227f29b989d0a2eda0c59909"
**********
**********Change-Id: I19bce9795e1d7e88119f9c16bc5213192080513f"
*********
*********Change-Id: Ie1b4f393df88a6d3b6ee83050ab6ef316c00b1e3"""""""""
---
 linker/linker.cpp      |  8 ++++----
 linker/linker_main.cpp |  5 ++++-
 linker/shims.h         | 47 +++++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 55 insertions(+), 5 deletions(-)
 create mode 100644 linker/shims.h

diff --git a/linker/linker.cpp b/linker/linker.cpp
index 3d6fdf9..a4580e1 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -1046,8 +1046,8 @@ static int open_library_in_zipfile(ZipArchiveCache* zip_archive_cache,
   if (realpath_fd(fd, realpath)) {
     *realpath += separator;
   } else {
-    PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.",
-          normalized_path.c_str());
+    //PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.",
+     //     normalized_path.c_str());
     *realpath = normalized_path;
   }
 
@@ -1084,7 +1084,7 @@ static int open_library_on_paths(ZipArchiveCache* zip_archive_cache,
       if (fd != -1) {
         *file_offset = 0;
         if (!realpath_fd(fd, realpath)) {
-          PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.", buf);
+          //PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.", buf);
           *realpath = buf;
         }
       }
@@ -1117,7 +1117,7 @@ static int open_library(android_namespace_t* ns,
       if (fd != -1) {
         *file_offset = 0;
         if (!realpath_fd(fd, realpath)) {
-          PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.", name);
+          //PRINT("warning: unable to get realpath for the library \"%s\". Will use given path.", name);
           *realpath = name;
         }
       }
diff --git a/linker/linker_main.cpp b/linker/linker_main.cpp
index ac73920..24395fe 100644
--- a/linker/linker_main.cpp
+++ b/linker/linker_main.cpp
@@ -47,6 +47,8 @@
 
 #include <vector>
 
+#include "shims.h"
+
 extern void __libc_init_globals(KernelArgumentBlock&);
 extern void __libc_init_AT_SECURE(KernelArgumentBlock&);
 
@@ -261,9 +263,10 @@ static ElfW(Addr) __linker_init_post_relocation(KernelArgumentBlock& args, ElfW(
     if (ldpreload_env != nullptr) {
       INFO("[ LD_PRELOAD set to \"%s\" ]", ldpreload_env);
     }
-    ldshim_libs_env = getenv("LD_SHIM_LIBS");
   }
 
+  ldshim_libs_env = LD_SHIM_LIBS;
+ 
   struct stat file_stat;
   // Stat "/proc/self/exe" instead of executable_path because
   // the executable could be unlinked by this point and it should
diff --git a/linker/shims.h b/linker/shims.h
new file mode 100644
index 0000000..1b5b49e
--- /dev/null
+++ b/linker/shims.h
@@ -0,0 +1,47 @@
+
+/*
+ * Copyright (C) 2017 Jonathan Dennis [Meticulus]
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef __SHIMS_H
+#define __SHIMS_H
+
+#define COLON ":"
+
+#define BERLIN_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.BERLIN.so|libshim_libgui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.BERLIN.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.BERLIN.so|libshim.so"
+#define DALLAS_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.DALLAS.so|libshim_libgui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.DALLAS.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.DALLAS.so|libshim.so"
+#define NEMO_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.NEMO.so|libshim_libgui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.NEMO.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.NEMO.so|libshim.so"
+#define PRAGUE_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.PRAGUE.so|libshim_libgui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.PRAGUE.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.PRAGUE.so|libshim.so"
+#define VENUS_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.VENUS.so|libshim_libgui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.VENUS.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.VENUS.so|libshim.so"
+#define WARSAW_SHIMS "/hwvendor/lib64/hwcam/hwcam.hi6250.m.WARSAW.so|libshim_libui.so:/hwvendor/lib64/hwcam/hwcam.hi6250.m.WARSAW.so|libshim.so"
+
+#define COMMON_SHIMS "/hwvendor/lib/hw/audio.primary.hisi.so|libshim.so:/hwvendor/lib64/libcamera_algo.so|libshim_libui.so:/hwvendor/lib64/hw/fingerprint.hi6250.so|libshim.so"
+
+#define OREO_SHIMS "/hwvendor/lib/hw/audio.primary.hisi.so|libshim_icu.so"
+
+#define LD_SHIM_LIBS COMMON_SHIMS COLON BERLIN_SHIMS COLON DALLAS_SHIMS COLON NEMO_SHIMS COLON PRAGUE_SHIMS COLON VENUS_SHIMS COLON WARSAW_SHIMS COLON OREO_SHIMS
+#endif
-- 
2.7.4

